azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import logging
+import uuid
+from threading import Timer
+from transitions import Machine
+from ..transport import constant
+
+from .action import SubscribeAction, PublishAction, UnsubscribeAction
+from .request_response_provider import RequestResponseMachine
+
+
+logger = logging.getLogger(__name__)
+
+
+class PollingMachine(object):
+    """
+    Class that is responsible for sending the initial registration request and polling the
+    registration process for constant updates.
+    """
+
+    def __init__(self, state_based_provider):
+        """
+        :param state_based_provider: The state machine based provider.
+        """
+        self._polling_timer = None
+        self._query_timer = None
+
+        self._register_callback = None
+        self._cancel_callback = None
+
+        self.on_connected = None
+        self.on_disconnected = None
+        self.on_registration_status_update = None
+
+        self._operations = {}
+
+        self._request_response_provider = RequestResponseMachine(state_based_provider)
+
+        self._request_response_provider.on_response_received = self._handle_response_received
+
+        states = [
+            "disconnected",
+            "initializing",
+            "registering",
+            "waiting_to_poll",
+            "polling",
+            "completed",
+            "error",
+            "cancelling",
+        ]
+
+        transitions = [
+            {
+                "trigger": "_trig_register",
+                "source": "disconnected",
+                "before": "_initialize_register",
+                "dest": "initializing",
+            },
+            {
+                "trigger": "_trig_register",
+                "source": "error",
```

Comment:
>error [](start = 27, length = 5)

how is "error" different from "disconnected" other than the name?  It seems to behave the same.  I think you need to consider what actions to take on error, if that means disconnecting, or staying in some "unrecoverable" state that requires the user to start over, or something else.  Right now, it seems ambiguous

Reply:
after defining the states ,  i felt it is behaving the same as cancelling. But initially error was defined as a state to transition into whenever the time is up or any other error happened. While cancelling is something that the user decides to do.

---
In reply to: [277411944](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277411944) [](ancestors = 277411944)

Reply:
Right.  You get into the error state a different way, but it looks like the machine behaves exactly the same if you are in "disconnected" or if you are in "error", which makes me think that "error" shouldn't be it's own state.  

---
In reply to: [277467697](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277467697) [](ancestors = 277467697,277411944)

Reply:
right now the `_call_error` and `_call_complete` looks same to me. During `trig_error` I dont see that I have called unsubscribe or disconnect. If we are doing unsubscribe and disconnect when we run into an error then the error and cancel state would look same. 

---
In reply to: [277792329](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277792329) [](ancestors = 277792329,277467697,277411944)

Reply:
the difference is the result that is getting passed as a param into the callback, merging this would require an instance check

---
In reply to: [278317931](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r278317931) [](ancestors = 278317931,277792329,277467697,277411944)